jax/tree_util.py

```diff
@@ -193,13 +193,29 @@ def build_tree(treedef, xs):
   return treedef.from_iterable_tree(xs)
 
 def tree_transpose(outer_treedef, inner_treedef, pytree_to_transpose):
-  flat, treedef = tree_flatten(pytree_to_transpose)
-  expected_treedef = outer_treedef.compose(inner_treedef)
-  if treedef != expected_treedef:
-    raise TypeError("Mismatch\n{}\n != \n{}".format(treedef, expected_treedef))
+  """Produces a pytree with the same data but a different structure.
```

Comment:
I am proposing that we add some documentation to tree_transpose (I did not know what it did)